### ISSUE TITLE
Unhelpful static_assert pointing to a TODO in traced_value.h

### ISSUE THREAD

**Issue by rburchell** [google/perfetto#375]
Trying to add some tracing, I get a compile error:

> error: static assertion failed: The provided type (passed to TRACE_EVENT argument / TracedArray::Append / TracedDictionary::Add) does not support being written in a trace format. Please see the comment in traced_value.h for more details.

Okay, fair enough. I'm sure that I'll find some useful assistance with my problems in the header ...

```cpp
void WriteIntoTracedValue(TracedValue context, T&& value) {
  // TODO(altimin): Add a URL to documentation and a list of common failure
  // patterns.
  static_assert(
      internal::has_traced_value_support<T>::value,
      "The provided type (passed to TRACE_EVENT argument / TracedArray::Append "
      "/ TracedDictionary::Add) does not support being written in a trace "
      "format. Please see the comment in traced_value.h for more details.");
```

... or maybe not.

I'd suggest that either that TODO gets TODONE, or (less effort), you remove the final sentence to avoid frustrating the next person to hit it, since it doesn't seem to be especially high priority, since it's still there after being added two years ago.

**Comment by altimin** [comment]
Hey, thanks for the report!

The intention of this comment was to point to [the TracedValue's class-level comment](https://cs.android.com/android/platform/superproject/+/master:external/perfetto/include/perfetto/tracing/traced_value.h;l=44;drc=22739b488b8f270bc50088e6d255732880465ba5).

I also started looking into [updating the docs](https://android-review.googlesource.com/c/platform/external/perfetto/+/2031943) earlier this year, but dropped a ball there – sorry about this, will try to get to it soon.

**Comment by rburchell** [comment]
No problem, I understand that these things happen, but the guidance as it is isn't terribly clear :-)

But it sounds like you have it under control, so I'm fine with closing this if you prefer.

In my case, the problem was that I was passing a perfetto::Track() to TRACE_EVENT_END() "by accident", in that I wrote:

> TRACE_EVENT_BEGIN("foo", "bar", perfetto::Track(1234));
> TRACE_EVENT_END("foo", "bar", perfetto::Track(1234)); // oops, the "bar" shouldn't have been here

... so the documentation there (which I did read next) didn't actually help me, I had to step back and look at TRACE_EVENT_END to figure out where I was going wrong.

**Comment by altimin** [comment]
Yeah, the extra event name in `TRACE_EVENT_END` is a bit of a landmine and I see folks running into this regularly.

Fixing this seems hard without breaking backwards compatibility, but +@betasheet for thoughts here.

**Comment by betasheet** [comment]
Some thoughts/discussion on the TRACE_EVENT_END name issue are on crbug.com/1349601. One thing we could try is to add back the name under a compile time flag (gn arg or similar). We'd could eventually also try to make that the default flag (and allow setting a build time flag to opt into the old mechanism), but I'm not sure if that's going to be feasible?
